code/modules/mob/living/carbon/xenomorph/abilities.dm

```diff
@@ -241,6 +344,52 @@
 
 	return (base_wait + scaling_wait - max(0, (scaling_wait * X.health / X.maxHealth))) * build_resin_modifier
 
+/datum/action/xeno_action/activable/secrete_resin/proc/preshutter_build_resin(turf/T, silent = FALSE)
```

Comment:
A lot of code duplication as well.

That is why you need to refactor the action